generators/csharp/codegen/src/ast/Class.ts

```diff
@@ -353,27 +372,40 @@ export class Class extends AstNode {
 
     private writeOperator({ writer, operator }: { writer: Writer; operator: Class.Operator }): void {
         writer.write("public static ");
-        switch (operator.type) {
-            case "implicit":
-                writer.write("implicit ");
-                break;
-            case "explicit":
-                writer.write("explicit ");
-                break;
-            default:
-                assertNever(operator.type);
+        if (operator.type === "explicit" || operator.type === "implicit") {
+            switch (operator.type) {
+                case "implicit":
+                    writer.write("implicit ");
+                    break;
+                case "explicit":
+                    writer.write("explicit ");
+                    break;
+                default:
+                    assertNever(operator.type);
+            }
+            writer.write("operator ");
+            writer.write(`${this.name}(`);
+        } else {
+            if (!operator.return) {
+                throw new Error("Operator must have a return type, unless it's implicit or explicit");
+            }
```

Comment:
We should avoid throwing errors within the AST nodes - I think we can refactor things so that we have a separate `Operator` interface that actually requires the `return` parameter.

Given that this isn't user-facing, we can handle this in a follow-up.

Reply:
How does this look to you? I'm not a fan of the `NormalOperator` name.

```ts
    interface CastOperator {
        parameter: Parameter;
        type: "implicit" | "explicit";
        body: CodeBlock;
        useExpressionBody?: boolean;
    }
    interface NormalOperator {
        type: "==" | "!=";
        parameters: Parameter[];
        return: Type;
        body: CodeBlock;
        useExpressionBody?: boolean;
    }
    type Operator = CastOperator | NormalOperator;

    private writeOperator({ writer, operator }: { writer: Writer; operator: Class.Operator }): void {
        writer.write("public static ");
        if (operator.type === "explicit" || operator.type === "implicit") {
            writer.write(`${operator.type} `);
            writer.write("operator ");
            writer.write(`${this.name}(`);
            operator.parameter.write(writer);
        } else {
            const normalOperator = operator as Class.NormalOperator;
            normalOperator.return.write(writer);
            writer.write(" operator ");
            writer.write(`${operator.type}(`);
            normalOperator.parameters.forEach((parameter, idx) => {
                parameter.write(writer);
                if (idx < normalOperator.parameters.length - 1) {
                    writer.write(", ");
                }
            });
        }
        if (operator.useExpressionBody) {
            writer.write(") => ");
            writer.writeNodeStatement(operator.body);
        } else {
            writer.write(") {");
            writer.writeNode(operator.body);
            writer.writeLine("}");
        }
    }
```